.github/workflows/build-test-e2e.yml

```diff
@@ -1,11 +1,34 @@
 name: 'Build/Run End-to-End Tests'
 
 on: push
+env:
+  BUILD_ARTIFACT_NAME: swc-war
+  BUILD_DIR: target
+  BUILD_NAME: portal-develop-SNAPSHOT.war
+  NODE_VERSION: 18.16.0
+  PW_ALL_BLOBS_DIR: all-blob-reports
 jobs:
-  build-and-run-e2e:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Build SWC
+        uses: ./.github/workflows/build
+      - name: Upload build to GitHub Actions Artifacts
+        uses: actions/upload-artifact@v3
+        with:
+          name: ${{ env.BUILD_ARTIFACT_NAME }}
+          path: '${{ env.BUILD_DIR }}/${{ env.BUILD_NAME }}'
+          retention-days: 1
+  playwright-tests:
     if: ${{ github.ref_name == 'develop' || startsWith(github.ref_name, 'release-') || github.actor == github.repository_owner }}
+    needs: [build]
     runs-on: ${{ github.repository_owner == 'Sage-Bionetworks' && 'ubuntu-22.04-4core-16GBRAM-150GBSSD' || 'macos-latest' }}
     timeout-minutes: 60
+    strategy:
+      fail-fast: false
+      matrix:
+        shard: [1/3, 2/3, 3/3]
```

Comment:
Shard e2e tests for each browser across machines. Set [`fail-fast`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast) to false, so that all shards will finish running tests, even if tests on one shard fail.

Reply:
Just curious, is it smart enough to shard by browser engine?

Reply:
Good question! I don't see their sharding strategy explicitly defined in the docs, but from my tests, sharding is done by test, where tests are ordered by the project declaration order in the config file, then the alphabetized file name, then the test declaration order within each file. The tests are then distributed as evenly as possible across shards. In a simple example, running three projects (10 chromium tests, 10 firefox tests, and 5 webkit tests) on three shards results in shard 1 running 9 tests (all chromium), shard 2 running 8 tests (1 chromium, 7 firefox), and shard 3 running 8 tests (3 firefox, 5 webkit).

Since we have one project per browser that runs all tests, using an equal number of shards as browsers ensures that each shard only runs tests for one browser. However, if we start writing browser-specific tests or change how the projects are defined, we may need a more complicated strategy to ensure that each shard runs one browser (or decide whether that's necessary).

I've seen issues requesting other sharding strategies (e.g. sharding by test time), so it's also possible that more sophisticated strategies will be available in the future.